src/QirRuntime/experimental/IonqTranslator.cpp

```diff
@@ -0,0 +1,665 @@
+
+    /*==============================================================================
+        As the simulator executes along a particular path, the translator would
+        accumulate corresponding instructions and errors.
+    ==============================================================================*/
+    struct TranslationError
+    {
+        Gate gate = Gate::undefined;
+        vector<long> targets;  // can be empty if error isn't specific to a particular qubit
+        vector<long> controls; // can be empty
+        string msg;
+
+        void Print(ostringstream& os) const
+        {
+            os << "{";
+            os << "\"gate\":\"" << gate_names[static_cast<size_t>(gate)] << "\",";
+            PrintVector(os, targets, "target");
+            PrintVector(os, controls, "control");
+            os << "\"message\":\"" << msg << "\"";
+            os << "},";
+        }
+    };
+
+    struct Instruction
+    {
+        Gate gate;
+        vector<long> targets;
+        vector<long> controls; // can be empty
+        double rotation = 0;   // only meaningful for rx, ry, rz, xx, yy and zz gates
+
+        Instruction(Gate gate, vector<long>&& targets, vector<long>&& controls, double rotation)
+            : gate(gate)
+            , targets(std::move(targets))
+            , controls(std::move(controls))
+            , rotation(rotation)
+        {
+        }
+
+        bool IsRotation() const
+        {
+            return Gate::rx <= gate && gate <= Gate::zz;
+        }
+
+        void Print(ostringstream& os) const
+        {
+            os << "{";
+            os << "\"gate\":\"" << gate_names[static_cast<size_t>(gate)] << "\",";
+            PrintVector(os, targets, "target");
+            PrintVector(os, controls, "control");
+            if (IsRotation())
+            {
+                os << "\"rotation\":" << rotation << ",";
+            }
+            os << "},";
+        }
+    };
+
+    /*==============================================================================
+        The decomposer provides classical "text book" decompositions for some of the
+        unitaries without any hardware specific optimizations.
+
+        A quantum processor, that wants to avail of the decompositions, should
+        implement the primitive operations, pass itself to the decomposer and then
+        delegate to it calls that should be decomposed.
+
+        NB: because the owner opts-in for each decomposition separately the decomposer
+        will call back for each step, even if it might come back to it (R1Frac -> R1).
+    ==============================================================================*/
+    class CCanonicalDecomposer final : public CQuantumApiBase
+    {
+        // The decomposer assumes that owner's lifetime exceeds its own.
+        IQuantumApi* owner = nullptr;
+
+        static double CalculateAngle(long numerator, long power)
+        {
+            return -numerator * M_PI / std::pow(2.0, power - 1);

```

Comment:
Looks like this produces a warning-as-error on the newer clang-tidy version:
`error: narrowing conversion from 'long' to 'double' [bugprone-narrowing-conversions,-warnings-as-errors]`. The same error crops up on line 226 as well.

Reply:
Changed `long power` to `int power` throughout the interfaces because we don't expect the clients to need super-large power values and providing support for them isn't trivial.